### PR TITLE
feat(NoteToSelf): add tasks counter

### LINKED ISSUE #13034: Add a counter for checkboxes with a "To Do" title in "Note to self" chat
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Feature request

**Which Nextcloud Version are you currently using:** NC30

**Is your feature request related to a problem? Please describe.**
When using the "Notes to self" chat as a flexible To Do list, the chat itself might hide some checkboxes above.

Will only make sense if https://github.com/nextcloud/spreed/issues/13033 is implemented.

**Describe the solution you'd like**
Similar to Github, a counter for unchecked checkboxes in the "Note to self" chat would be cool.

**Describe alternatives you've considered**
None

**Additional context**
Github:

image

### PR DESCRIPTION
**Pull request by DorraJaouad**
### ☑️ Resolves

* Fix #13034

### 🖼️ Screenshots / Screencasts

| Nothing done | Partially completed | 🏡 Completed
-- | -- | --
image | image | image

### 🏁 Checklist

- [ ] 🌏 Tested with different browsers / clients:
  - [x] Chromium (Chrome / Edge / Opera / Brave)
  - [ ] Firefox
  - [ ] Safari
  - [ ] Talk Desktop
  - [ ] Not risky to browser differences / client
- [ ] 🖌️ Design was reviewed, approved or inspired by the design team
- [ ] ⛑️ Tests are included or not possible
- [ ] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required

### CODE REVIEW

File: src/components/MessagesList/MessagesList.vue [hunk 1]
```diff
@@ -1238,6 +1245,15 @@ export default {
 			// scroll down by the height difference
 			this.$refs.scroller.scrollTop += heightDiff
 		},
+
+		updateTasksCount() {
+			if (!this.$refs.scroller) {
+				return
+			}
+			const tasksDoneCount = this.$refs.scroller.querySelectorAll('.checkbox-content__icon--checked')?.length
```
Comment (Antreesy) [inline src/components/MessagesList/MessagesList.vue:1253]: Another way could be to check message store for messages (`- [ ] ` and `- [x] `?

In case there would be an opportunity to render not all messages (virtual scroller)

File: src/components/TopBar/TasksCounter.vue [hunk 2]
```diff
@@ -0,0 +1,83 @@
+<!--
+  - SPDX-FileCopyrightText: 2023 Nextcloud GmbH and Nextcloud contributors
+  - SPDX-License-Identifier: AGPL-3.0-or-later
+-->
+
+<template>
+	<div class="tasks-counter">
+		<NcProgressBar type="circular" :value="tasksRatio" :color="tasksRatio === 100 ? 'green' : null" />
```
Comment (Antreesy) [inline src/components/TopBar/TasksCounter.vue:8]: var(--color-success)

**Review by Antreesy — COMMENTED** [review]
Otherwise - looks neat!

File: src/components/TopBar/TasksCounter.vue [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+		<div class="tasks-counter__count">
+			{{ tasksSummary }}
+		</div>
+	</div>
+</template>
+
+<script>
```
[the remark this answers is not in the document]
Comment (DorraJaouad) [reply src/components/TopBar/TasksCounter.vue:15]: Yes, I had this idea in mind and then I forgot 🙊

File: src/components/MessagesList/MessagesList.vue (hunk 1 above)
Comment (DorraJaouad) [reply src/components/MessagesList/MessagesList.vue:1253]: Performance-wise, it is more expensive as you will loop through messages.
Accuracy-wise, not all `- [ ]` are considered a checklist, maybe we need to do the same check when clicking on a list item to change the state.

**Review by Antreesy — APPROVED** [review]

**Comment by Antreesy** [comment]
@nickvergessen shall we backport this now, or respecting the string freeze - only after the 20.0 release?

File: src/components/TopBar/TasksCounter.vue [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	}
+	return (tasksDoneCount.value / tasksCount.value) * 100
+})
+
+const tasksSummary = computed(() => {
+	if (tasksRatio.value === 100) {
+		return t('spreed', 'All tasks done!')
+	}
+	// TRANSLATORS number of tasks done of total number of tasks
+	return t('spreed', '{done} of {total} tasks', {
```
Comment (nickvergessen) [inline src/components/TopBar/TasksCounter.vue:32]: I think this needs plurals?

**Comment by nickvergessen** [comment]
Drawn apart. I'd say lets do it, but the translation string looks "wrong" anyway

File: src/components/TopBar/TopBar.vue [hunk 5]
```diff
@@ -299,6 +305,7 @@ export default {
 
 		showUpcomingEvent() {
 			return this.nextEvent && !this.isInCall && !this.isSidebar && !this.isMobile
+				&& this.conversation.type === CONVERSATION.TYPE.NOTE_TO_SELF
 		},
```
Comment (ShGKme) [inline src/components/TopBar/TopBar.vue:309]: Seems to be unrelated and breaking upcoming events @DorraJaouad

Comment (Antreesy) [reply src/components/TopBar/TopBar.vue:309]: Or should be `!==`

Comment (ShGKme) [reply src/components/TopBar/TopBar.vue:309]: > Or supposed to be `!==`

Still not related to task counting though 👀

Comment (DorraJaouad) [reply src/components/TopBar/TopBar.vue:309]: Sorry, it should be `!==` indeed

Comment (DorraJaouad) [reply src/components/TopBar/TopBar.vue:309]: > > Or supposed to be `!==`
> 
> Still not related to task counting though 👀

Also true, but as I was adding the counter and touching that space in `TopBar`, I thought it could be slipped in this PR.
